File: runners/python/steps/response.py
from behave import given, when, then
from assertpy import assert_that
import json

# Advanced API assertion step definitions for Python SDK

# ReadChanges API assertions
@then('the response should contain "{field_name}"')
def step_response_should_contain_field(context, field_name):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    if field_name == 'changes':
        if not hasattr(context.test_context.last_response, 'changes') or context.test_context.last_response.changes is None:
            raise AssertionError('Response does not contain "changes" field')
    elif field_name == 'objects':
        if not hasattr(context.test_context.last_response, 'objects') or context.test_context.last_response.objects is None:
            raise AssertionError('Response does not contain "objects" field')
    else:
        if not hasattr(context.test_context.last_response, field_name):
            raise AssertionError(f'Response does not contain "{field_name}" field')

@then('the response should contain at most {max_count:d} changes')
def step_response_should_contain_at_most_changes(context, max_count):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response or not hasattr(context.test_context.last_response, 'changes'):
        raise AssertionError('Response does not contain changes field')

    actual_count = len(context.test_context.last_response.changes)
    if actual_count > max_count:
        raise AssertionError(f'Expected at most {max_count} changes, got {actual_count}')

@then('the response should have continuation token')
def step_response_should_have_continuation_token(context):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    if not hasattr(context.test_context.last_response, 'continuation_token') or not context.test_context.last_response.continuation_token:
        raise AssertionError('Response does not have continuation token')

@then('the response should not have continuation token')
def step_response_should_not_have_continuation_token(context):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    if hasattr(context.test_context.last_response, 'continuation_token') and context.test_context.last_response.continuation_token:
        raise AssertionError('Response has continuation token but should not have one')

@then('the changes should be different from "{saved_key}"')
def step_changes_should_be_different_from(context, saved_key):
    context.test_context.assert_response_success()
    
    if not hasattr(context.test_context, 'saved_data') or saved_key not in context.test_context.saved_data:
        raise AssertionError(f'No saved data found for key: {saved_key}')

    saved_response = context.test_context.saved_data[saved_key]
    
    # Simple comparison - in real implementation, would compare actual change content
    current_changes = getattr(context.test_context.last_response, 'changes', [])
    saved_changes = getattr(saved_response.get('response'), 'changes', [])
    
    if json.dumps(current_changes, sort_keys=True) == json.dumps(saved_changes, sort_keys=True):
        raise AssertionError(f'Changes are identical to saved data from {saved_key}')

@then('each change should have type "{expected_type}"')
def step_each_change_should_have_type(context, expected_type):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response or not hasattr(context.test_context.last_response, 'changes'):
        raise AssertionError('Response does not contain valid changes field')

    changes = context.test_context.last_response.changes
    for i, change in enumerate(changes):
        if not hasattr(change, 'type') or change.type != expected_type:
            actual_type = getattr(change, 'type', None)
            raise AssertionError(f'Change at index {i} has type "{actual_type}", expected "{expected_type}"')

# ListObjects API assertions
@then('the response should contain exactly {expected_count:d} objects')
def step_response_should_contain_exactly_objects(context, expected_count):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response or not hasattr(context.test_context.last_response, 'objects'):
        raise AssertionError('Response does not contain valid objects field')

    actual_count = len(context.test_context.last_response.objects)
    if actual_count != expected_count:
        raise AssertionError(f'Expected exactly {expected_count} objects, got {actual_count}')

@then('the response should contain exactly {expected_count:d} changes')
def step_response_should_contain_exactly_changes(context, expected_count):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response or not hasattr(context.test_context.last_response, 'changes'):
        raise AssertionError('Response does not contain valid changes field')

    actual_count = len(context.test_context.last_response.changes)
    if actual_count != expected_count:
        raise AssertionError(f'Expected exactly {expected_count} changes, got {actual_count}')

# Streaming API assertions
@then('the streaming response should be successful')
def step_streaming_response_should_be_successful(context):
    if context.test_context.last_error:
        raise AssertionError(f'Streaming response failed: {context.test_context.last_error}')

    if not context.test_context.last_response:
        raise AssertionError('No streaming response received')

@then('the streaming response should contain objects')
def step_streaming_response_should_contain_objects(context):
    context.test_context.assert_response_success()
    
    # For streaming responses, check if we received any objects
    if not hasattr(context.test_context, 'streamed_objects') or not context.test_context.streamed_objects:
        raise AssertionError('Streaming response did not contain any objects')

@then('each streamed object should have required fields')
def step_each_streamed_object_should_have_required_fields(context):
    context.test_context.assert_response_success()
    
    if not hasattr(context.test_context, 'streamed_objects') or not context.test_context.streamed_objects:
        raise AssertionError('No streamed objects to validate')

    for i, obj in enumerate(context.test_context.streamed_objects):
        if not hasattr(obj, 'object') or not obj.object:
            raise AssertionError(f'Streamed object at index {i} does not have required "object" field')

@then('the streaming connection should be properly closed')
def step_streaming_connection_should_be_properly_closed(context):
    # Placeholder for streaming connection validation
    # In real implementation, would check connection state
    if hasattr(context.test_context, 'streaming_connection') and context.test_context.streaming_connection:
        if getattr(context.test_context.streaming_connection, 'closed', False) is False:
            raise AssertionError('Streaming connection was not properly closed')

# Advanced response validation
@then('I save the response as "{save_key}"')
def step_save_response_as(context, save_key):
    context.test_context.assert_response_success()
    
    if not hasattr(context.test_context, 'saved_data'):
        context.test_context.saved_data = {}
    
    context.test_context.saved_data[save_key] = {
        'response': context.test_context.last_response,
        'timestamp': context.test_context.get_current_timestamp()
    }

@then('the response should contain at least {min_count:d} items')
def step_response_should_contain_at_least_items(context, min_count):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    item_count = 0
    
    # Check for different possible item fields
    if hasattr(context.test_context.last_response, 'changes') and context.test_context.last_response.changes:
        item_count = len(context.test_context.last_response.changes)
    elif hasattr(context.test_context.last_response, 'objects') and context.test_context.last_response.objects:
        item_count = len(context.test_context.last_response.objects)
    elif hasattr(context.test_context.last_response, 'tuples') and context.test_context.last_response.tuples:
        item_count = len(context.test_context.last_response.tuples)
    else:
        raise AssertionError('Response does not contain countable items (changes, objects, or tuples)')

    if item_count < min_count:
        raise AssertionError(f'Expected at least {min_count} items, got {item_count}')

@then('the response should have field "{field_name}"')
def step_response_should_have_field(context, field_name):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    if not hasattr(context.test_context.last_response, field_name):
        raise AssertionError(f'Response does not have field "{field_name}"')

@then('the response field "{field_name}" should be "{expected_value}"')
def step_response_field_should_be(context, field_name, expected_value):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    if not hasattr(context.test_context.last_response, field_name):
        raise AssertionError(f'Response does not have field "{field_name}"')

    actual_value = getattr(context.test_context.last_response, field_name)
    if str(actual_value) != expected_value:
        raise AssertionError(f'Response field "{field_name}" expected "{expected_value}", got "{actual_value}"')

# Pagination assertions
@then('the response should have pagination info')
def step_response_should_have_pagination_info(context):
    context.test_context.assert_response_success()
    
    if not context.test_context.last_response:
        raise AssertionError('No response received')

    # Check for common pagination fields
    has_continuation_token = hasattr(context.test_context.last_response, 'continuation_token') and context.test_context.last_response.continuation_token
    has_page_size = hasattr(context.test_context.last_response, 'page_size')
    has_next_page = hasattr(context.test_context.last_response, 'has_next_page')

    if not (has_continuation_token or has_page_size or has_next_page):
        raise AssertionError('Response does not contain pagination information')

# Error scenario assertions
@then('the error should contain "{expected_message}"')
def step_error_should_contain(context, expected_message):
    if not context.test_context.last_error:
        raise AssertionError('Expected an error but none occurred')

    error_message = str(context.test_context.last_error)
    if expected_message not in error_message:
        raise AssertionError(f'Error message "{error_message}" does not contain "{expected_message}"')

@then('the error should be of type "{expected_type}"')
def step_error_should_be_of_type(context, expected_type):
    if not context.test_context.last_error:
        raise AssertionError('Expected an error but none occurred')

    error_type = type(context.test_context.last_error).__name__
    if error_type != expected_type:
        raise AssertionError(f'Error type "{error_type}" does not match expected "{expected_type}"')

# Multi-response validation
@then('all responses should be successful')
def step_all_responses_should_be_successful(context):
    if not hasattr(context.test_context, 'response_history') or not context.test_context.response_history:
        raise AssertionError('No response history available')

    for i, record in enumerate(context.test_context.response_history):
        if record.get('error'):
            raise AssertionError(f'Response {i + 1} failed: {record["error"]}')

@then('the last {count:d} responses should be successful')
def step_last_responses_should_be_successful(context, count):
    if not hasattr(context.test_context, 'response_history') or len(context.test_context.response_history) < count:
        available = len(context.test_context.response_history) if hasattr(context.test_context, 'response_history') else 0
        raise AssertionError(f'Expected at least {count} responses, got {available}')

    last_responses = context.test_context.response_history[-count:]
    for i, record in enumerate(last_responses):
        if record.get('error'):
            raise AssertionError(f'Response {i + 1} of last {count} failed: {record["error"]}')

# Helper functions for advanced assertions
def validate_response_structure(response, expected_structure):
    """Validate response has expected structure"""
    for field_name, expected_type in expected_structure.items():
        if not hasattr(response, field_name):
            raise AssertionError(f'Response missing required field: {field_name}')
        
        actual_value = getattr(response, field_name)
        if expected_type != 'any':
            actual_type = type(actual_value).__name__
            if actual_type != expected_type:
                raise AssertionError(f'Response field "{field_name}" expected type "{expected_type}", got "{actual_type}"')

def get_response_field(response, field_path):
    """Get nested field from response using dot notation"""
    if not response:
        return None
    
    parts = field_path.split('.')
    current = response
    
    for part in parts:
        if current is None:
            return None
        if not hasattr(current, part):
            return None
        current = getattr(current, part)
    
    return current

def compare_responses(response1, response2, ignore_fields=None):
    """Compare two responses, optionally ignoring certain fields"""
    if ignore_fields is None:
        ignore_fields = []
    
    # Convert responses to dictionaries for comparison
    dict1 = response_to_dict(response1, ignore_fields)
    dict2 = response_to_dict(response2, ignore_fields)
    
    return json.dumps(dict1, sort_keys=True) == json.dumps(dict2, sort_keys=True)

def response_to_dict(response, ignore_fields=None):
    """Convert response object to dictionary, ignoring specified fields"""
    if ignore_fields is None:
        ignore_fields = []
    
    result = {}
    for attr in dir(response):
        if not attr.startswith('_') and attr not in ignore_fields:
            value = getattr(response, attr)
            if not callable(value):
                result[attr] = value
    
    return result

def store_response_in_history(test_context, response, error=None):
    """Store response in history for multi-response validation"""
    if not hasattr(test_context, 'response_history'):
        test_context.response_history = []
    
    test_context.response_history.append({
        'response': response,
        'error': error,
        'timestamp': test_context.get_current_timestamp()
    })

File: runners/js/features/step_definitions/response.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { expect } from 'chai';
import { TestWorld } from './world';

// Advanced API assertion step definitions for JavaScript SDK

// ReadChanges API assertions
Then('the response should contain {string}', function(this: TestWorld, fieldName: string) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  if (fieldName === 'changes') {
    if (!this.lastResponse.changes) {
      throw new Error('Response does not contain "changes" field');
    }
  } else if (fieldName === 'objects') {
    if (!this.lastResponse.objects) {
      throw new Error('Response does not contain "objects" field');
    }
  } else {
    if (!this.lastResponse[fieldName]) {
      throw new Error(`Response does not contain "${fieldName}" field`);
    }
  }
});

Then('the response should contain at most {int} changes', function(this: TestWorld, maxCount: number) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse || !this.lastResponse.changes) {
    throw new Error('Response does not contain changes field');
  }

  const actualCount = this.lastResponse.changes.length;
  if (actualCount > maxCount) {
    throw new Error(`Expected at most ${maxCount} changes, got ${actualCount}`);
  }
});

Then('the response should have continuation token', function(this: TestWorld) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  if (!this.lastResponse.continuation_token) {
    throw new Error('Response does not have continuation token');
  }
});

Then('the response should not have continuation token', function(this: TestWorld) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  if (this.lastResponse.continuation_token) {
    throw new Error('Response has continuation token but should not have one');
  }
});

Then('the changes should be different from {string}', function(this: TestWorld, savedKey: string) {
  this.assertResponseSuccess();
  
  if (!this.savedData || !this.savedData.has(savedKey)) {
    throw new Error(`No saved data found for key: ${savedKey}`);
  }

  const savedResponse = this.savedData.get(savedKey);
  
  // Simple comparison - in real implementation, would compare actual change content
  if (JSON.stringify(this.lastResponse.changes) === JSON.stringify(savedResponse.changes)) {
    throw new Error(`Changes are identical to saved data from ${savedKey}`);
  }
});

Then('each change should have type {string}', function(this: TestWorld, expectedType: string) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse || !this.lastResponse.changes) {
    throw new Error('Response does not contain valid changes field');
  }

  for (let i = 0; i < this.lastResponse.changes.length; i++) {
    const change = this.lastResponse.changes[i];
    if (!change.type || change.type !== expectedType) {
      throw new Error(`Change at index ${i} has type '${change.type}', expected '${expectedType}'`);
    }
  }
});

// ListObjects API assertions
Then('the response should contain exactly {int} objects', function(this: TestWorld, expectedCount: number) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse || !this.lastResponse.objects) {
    throw new Error('Response does not contain valid objects field');
  }

  const actualCount = this.lastResponse.objects.length;
  if (actualCount !== expectedCount) {
    throw new Error(`Expected exactly ${expectedCount} objects, got ${actualCount}`);
  }
});

Then('the response should contain exactly {int} changes', function(this: TestWorld, expectedCount: number) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse || !this.lastResponse.changes) {
    throw new Error('Response does not contain valid changes field');
  }

  const actualCount = this.lastResponse.changes.length;
  if (actualCount !== expectedCount) {
    throw new Error(`Expected exactly ${expectedCount} changes, got ${actualCount}`);
  }
});

// Streaming API assertions
Then('the streaming response should be successful', function(this: TestWorld) {
  if (this.lastError) {
    throw new Error(`Streaming response failed: ${this.lastError.message}`);
  }

  if (!this.lastResponse) {
    throw new Error('No streaming response received');
  }
});

Then('the streaming response should contain objects', function(this: TestWorld) {
  this.assertResponseSuccess();
  
  // For streaming responses, check if we received any objects
  if (!this.streamedObjects || this.streamedObjects.length === 0) {
    throw new Error('Streaming response did not contain any objects');
  }
});

Then('each streamed object should have required fields', function(this: TestWorld) {
  this.assertResponseSuccess();
  
  if (!this.streamedObjects || this.streamedObjects.length === 0) {
    throw new Error('No streamed objects to validate');
  }

  for (let i = 0; i < this.streamedObjects.length; i++) {
    const obj = this.streamedObjects[i];
    if (!obj.object) {
      throw new Error(`Streamed object at index ${i} does not have required 'object' field`);
    }
  }
});

Then('the streaming connection should be properly closed', function(this: TestWorld) {
  // Placeholder for streaming connection validation
  // In real implementation, would check connection state
  if (this.streamingConnection && this.streamingConnection.readyState !== 'closed') {
    throw new Error('Streaming connection was not properly closed');
  }
});

// Advanced response validation
Then('I save the response as {string}', function(this: TestWorld, saveKey: string) {
  this.assertResponseSuccess();
  
  this.savedData.set(saveKey, {
    response: this.lastResponse,
    timestamp: new Date().toISOString()
  });
});

Then('the response should contain at least {int} items', function(this: TestWorld, minCount: number) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  let itemCount = 0;
  
  // Check for different possible item fields
  if (this.lastResponse.changes) {
    itemCount = this.lastResponse.changes.length;
  } else if (this.lastResponse.objects) {
    itemCount = this.lastResponse.objects.length;
  } else if (this.lastResponse.tuples) {
    itemCount = this.lastResponse.tuples.length;
  } else {
    throw new Error('Response does not contain countable items (changes, objects, or tuples)');
  }

  if (itemCount < minCount) {
    throw new Error(`Expected at least ${minCount} items, got ${itemCount}`);
  }
});

Then('the response should have field {string}', function(this: TestWorld, fieldName: string) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  if (this.lastResponse[fieldName] === undefined) {
    throw new Error(`Response does not have field '${fieldName}'`);
  }
});

Then('the response field {string} should be {string}', function(this: TestWorld, fieldName: string, expectedValue: string) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  const actualValue = this.lastResponse[fieldName];
  if (actualValue === undefined) {
    throw new Error(`Response does not have field '${fieldName}'`);
  }

  if (actualValue.toString() !== expectedValue) {
    throw new Error(`Response field '${fieldName}' expected '${expectedValue}', got '${actualValue}'`);
  }
});

// Pagination assertions
Then('the response should have pagination info', function(this: TestWorld) {
  this.assertResponseSuccess();
  
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  // Check for common pagination fields
  const hasContinuationToken = this.lastResponse.continuation_token !== undefined;
  const hasPageSize = this.lastResponse.page_size !== undefined;
  const hasNextPage = this.lastResponse.has_next_page !== undefined;

  if (!hasContinuationToken && !hasPageSize && !hasNextPage) {
    throw new Error('Response does not contain pagination information');
  }
});

// Error scenario assertions
Then('the error should contain {string}', function(this: TestWorld, expectedMessage: string) {
  if (!this.lastError) {
    throw new Error('Expected an error but none occurred');
  }

  const errorMessage = this.lastError.message || this.lastError.toString();
  if (!errorMessage.includes(expectedMessage)) {
    throw new Error(`Error message '${errorMessage}' does not contain '${expectedMessage}'`);
  }
});

Then('the error should be of type {string}', function(this: TestWorld, expectedType: string) {
  if (!this.lastError) {
    throw new Error('Expected an error but none occurred');
  }

  const errorType = this.lastError.constructor.name || typeof this.lastError;
  if (errorType !== expectedType) {
    throw new Error(`Error type '${errorType}' does not match expected '${expectedType}'`);
  }
});

// Multi-response validation
Then('all responses should be successful', function(this: TestWorld) {
  if (!this.responseHistory || this.responseHistory.length === 0) {
    throw new Error('No response history available');
  }

  for (let i = 0; i < this.responseHistory.length; i++) {
    const record = this.responseHistory[i];
    if (record.error) {
      throw new Error(`Response ${i + 1} failed: ${record.error.message}`);
    }
  }
});

Then('the last {int} responses should be successful', function(count) {
  if (!this.responseHistory || this.responseHistory.length < count) {
    const available = this.responseHistory ? this.responseHistory.length : 0;
    throw new Error(`Expected at least ${count} responses, got ${available}`);
  }

  const lastResponses = this.responseHistory.slice(-count);
  for (let i = 0; i < lastResponses.length; i++) {
    const record = lastResponses[i];
    if (record.error) {
      throw new Error(`Response ${i + 1} of last ${count} failed: ${record.error.message}`);
    }
  }
});

// Helper methods for advanced assertions
function validateResponseStructure(response, expectedStructure) {
  for (const [key, type] of Object.entries(expectedStructure)) {
    if (response[key] === undefined) {
      throw new Error(`Response missing required field: ${key}`);
    }
    
    const actualType = typeof response[key];
    if (actualType !== type && type !== 'any') {
      throw new Error(`Response field '${key}' expected type '${type}', got '${actualType}'`);
    }
  }
}

// Add helper methods to World/TestContext
if (typeof module !== 'undefined' && module.exports) {
  const advancedAssertionHelpers = {
    validateResponseStructure,
    
    storeResponseInHistory: function(response, error) {
      if (!this.responseHistory) {
        this.responseHistory = [];
      }
      this.responseHistory.push({ response, error, timestamp: new Date().toISOString() });
    },

    getResponseField: function(fieldPath) {
      if (!this.lastResponse) return undefined;
      
      const parts = fieldPath.split('.');
      let current = this.lastResponse;
      
      for (const part of parts) {
        if (current === null || current === undefined) return undefined;
        current = current[part];
      }
      
      return current;
    },

    compareResponses: function(response1, response2, ignoreFields = []) {
      const clean1 = { ...response1 };
      const clean2 = { ...response2 };
      
      // Remove ignored fields
      for (const field of ignoreFields) {
        delete clean1[field];
        delete clean2[field];
      }
      
      return JSON.stringify(clean1) === JSON.stringify(clean2);
    }
  };

  module.exports = advancedAssertionHelpers;
}
